tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -139,8 +153,335 @@ impl ServerKey {
+                .step_by(i_plus_1_pow_2)
+                .map(|k| {
+                    // Since our carry_propagation LUT ie sum function
+                    // is not commutative we have to reverse operands
+                    self.key.unchecked_apply_lookup_table_bivariate(
+                        &carry_out[k + i_pow_2 - 1],
+                        &carry_out[k + i_plus_1_pow_2 - 1],
+                        &lut_carry_propagation_sum,
+                    )
+                })
+                .collect_into_vec(&mut buffer);
+
+            let mut drainer = buffer.drain(..);
+            for k in (0..num_blocks).step_by(i_plus_1_pow_2) {
+                let b = drainer.next().unwrap();
+                carry_out.swap(k + i_pow_2 - 1, k + i_plus_1_pow_2 - 1);
+                carry_out[k + i_plus_1_pow_2 - 1] = b;
+            }
+            drop(drainer);
+            assert!(buffer.is_empty());
+        }
+
+        // The first step of the Down-Sweep phase sets the
+        // first block to 0
+        lhs.blocks
+            .par_iter_mut()
+            .zip(carry_out.par_iter())
+            .for_each(|(block, carry_in)| {
+                self.key.unchecked_add_assign(block, carry_in);
+                self.key.message_extract_assign(block);
+            });
+    }
+
+    /// Initialization function for parallal carryless sum
+    ///
+    /// This function adds rhs into lhs
+    ///
+    /// It Returns the array that tells whether the
+    /// block resulting from the sum will propagat or generate
+    /// a carry. A Prefix sum / cumulative sum then needs to be done
+    /// to get the final carry that the block will output
+    fn add_and_generate_init_carry_array<PBSOrder: PBSOrderMarker>(
+        &self,
+        lhs: &mut RadixCiphertext<PBSOrder>,
+        rhs: &RadixCiphertext<PBSOrder>,
+        add_extra_one: AddExtraOne,
+    ) -> Vec<crate::shortint::CiphertextBase<PBSOrder>> {
+        let modulus = self.key.message_modulus.0 as u64;
+
+        let lut_does_block_generate_carry =
+            self.key.generate_accumulator(|x| (x >= modulus) as u64);
+
+        let lut_does_block_generate_or_propagate = self.key.generate_accumulator(|x| {
+            if x >= modulus {
+                OuputCarry::Generated as u64
+            } else if x == (modulus - 1) {
+                OuputCarry::Propagated as u64
+            } else {
+                OuputCarry::None as u64
+            }
+        });
+
+        let mut carry_out = Vec::with_capacity(lhs.blocks.len());
+        lhs.blocks
+            .par_iter_mut()
+            .zip(rhs.blocks.par_iter())
+            .enumerate()
+            .map(|(i, (ct_left_i, ct_right_i))| {
+                self.key.unchecked_add_assign(ct_left_i, ct_right_i);
+                if i == 0 {
+                    if matches!(add_extra_one, AddExtraOne::Yes) {
+                        self.key.unchecked_scalar_add_assign(ct_left_i, 1);
+                    }
```

Comment:
I'm guessing this is ok to do as lhs and rhs are assumed to have empty carries, correct ? If so worth a small comment here (and also indicate it's to simplify the sub algorithm if possible)